include/deal.II/numerics/vector_tools_constraints.templates.h

```diff
@@ -567,115 +578,122 @@ namespace VectorTools
-                    internal::VectorDoFTuple<dim> vector_dofs;
-                    vector_dofs.dof_indices[0] = face_dofs[i];
-
-                    Assert(
-                      first_vector_component + dim <= fe.n_components(),
-                      ExcMessage(
-                        "Error: the finite element does not have enough components "
-                        "to define a normal direction."));
-
-                    for (unsigned int k = 0; k < fe.n_dofs_per_face(face_no);
-                         ++k)
-                      if ((k != i) &&
-                          (face_quadrature_collection[cell->active_fe_index()]
-                             .point(k) ==
-                           face_quadrature_collection[cell->active_fe_index()]
-                             .point(i)) &&
-                          (fe.face_system_to_component_index(k, face_no)
-                             .first >= first_vector_component) &&
-                          (fe.face_system_to_component_index(k, face_no).first <
-                           first_vector_component + dim))
-                        vector_dofs.dof_indices
-                          [fe.face_system_to_component_index(k, face_no).first -
-                           first_vector_component] = face_dofs[k];
-
-                    for (unsigned int d = 0; d < dim; ++d)
-                      Assert(vector_dofs.dof_indices[d] < dof_handler.n_dofs(),
+                  // Refinement edge indices are going to be constrained to 0
+                  // during a multigrid cycle and do not need no-normal-flux
+                  // constraints, so skip them:
+                  if (!refinement_edge_indices.is_element(face_dofs[i]))
+                    {
+                      // find corresponding other components of vector
+                      internal::VectorDoFTuple<dim> vector_dofs;
+                      vector_dofs.dof_indices[0] = face_dofs[i];
+
+                      Assert(
+                        first_vector_component + dim <= fe.n_components(),
+                        ExcMessage(
+                          "Error: the finite element does not have enough components "
+                          "to define a normal direction."));
+
+                      for (unsigned int k = 0; k < fe.n_dofs_per_face(face_no);
+                           ++k)
+                        if ((k != i) &&
+                            (face_quadrature_collection[cell->active_fe_index()]
+                               .point(k) ==
+                             face_quadrature_collection[cell->active_fe_index()]
+                               .point(i)) &&
+                            (fe.face_system_to_component_index(k, face_no)
+                               .first >= first_vector_component) &&
+                            (fe.face_system_to_component_index(k, face_no)
+                               .first < first_vector_component + dim))
```

Comment:
This is not the fault of this PR, but this seems to be like a quadratic complexity in the number of dofs on a face. It should be possible to fix this by a lookup list outside the loop over the cells.

Reply:
Thanks, I will do it in a separate PR.